### ISSUE TITLE
CPU memory usage keeps on increasing during training

### ISSUE THREAD

**Issue by MLbyML** [mpicbg-csbd/stardist#107]
Hello,

Thanks for providing very clean notebooks.
I am using `StarDist3D` on some volumetric images. In my current experiment, I try to use 175 training images and 21 validation images. Each raw microscopy image is 21.2 MB large (and with dimensions z = 70 , x = 401, y = 378, 16 bit). I notice that even though I am using the GPU:
![nvidia-smi](https://user-images.githubusercontent.com/34229641/102105607-e7e71c00-3e2f-11eb-8583-6233e1f5c36a.png)
 
.... the CPU memory usage keeps on increasing over time (epochs). 

For example, 
prior to starting the notebook, running `free -m` gives: 
```
              total        used        free      shared  buff/cache   available
Mem:          30150         391       29290           9         469       29516
Swap:             0           0           0
```
at epoch = 0, running `free -m` gives:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       13564        7846          25        8739       16314
Swap:             0           0           0
```
and at epoch = 5, we are at:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       29305         270         159         574         443
Swap:             0           0           0
```

Soon after the notebook dies. 

Could there be some memory leak perhaps? 
Would you have any suggestions on how to avoid this problem and be able to use all the training data? 
Thank you for your time!

**Comment by maweigert** [comment]
Hi @MLbyML 

Indeed, that should not happend. What is the output of free -m

1) after 10 epochs?

2) when using only half of the training data?

Thanks for debugging! 

M

**Comment by MLbyML** [comment]
Hi @maweigert 
Thank you for the reply and the suggestions!

So already at epoch = 9, `free -m` gives the following while using all the training data:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       29401         266         159         482         349
Swap:             0           0           0
```
So it never reaches epoch = 10 because soon after all the memory is used and the kernel dies.

Regarding the second question, for my machine, using half the training data (`X = X[::2]; Y = Y[::2]`) works well although the CPU memory still initially (between epoch = 0 to epoch = 7) increases to some extent and then later reaches a steady state:

at epoch = 0, with half the training data:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       10692       10377         159        9080       19050
Swap:             0           0           0
```
at epoch = 5, with half the training data:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       18466        2551         159        9132       11275
Swap:             0           0           0
```

at epoch = 10, with half the training data:
```
              total        used        free      shared  buff/cache   available
Mem:          30150       20562         453         159        9134        9179
Swap:             0           0           0```
```

beyond this, it hovers around this (~20 GB) memory limit and doesn't increase significantly more.
Using half the training data would help. But it would also be useful for us to use of it. 
Happy to hear if you have any suggestions or would like me to debug anymore. 
Thank you for your time!

**Comment by uschmidt83** [comment]
Hi Manan,

having only 32 GB of RAM can be limiting when training with relatively large 3D datasets.
Your training data alone has (175+21) * (70*401*378) ~ 2 billion pixels, i.e. needs more than 8GB as (normalized) float32 images in memory.
Additionally, the StarDist data generator makes an internal copy to cache regions where it can sample patches from. There are probably a few other places where copies of the data are needed...

Best,
Uwe

**Comment by maweigert** [comment]
> Happy to hear if you have any suggestions or would like me to debug anymore.

You can as well avoid loading everything into memory altogether by using a `Sequence` that dynamically loads images and labels from disk (`model.train` should accepts a `Sequence` equally well)  

```python
from tensorflow.keras.utils import Sequence
import tifffile
 
class FileData(Sequence):
    def __init__(self, filenames):
        super().__init__() 
        self._filenames = filenames
    def __len__(self):
        return len(self._filenames)
    def __getitem__(self, i):
        x = tifffile.imread(self._filenames[i]) 
        # do something with it, e.g. normalize for raw images etc....
        return x

X = FileData(files_images)
Y = FileData(files_labels)

model.train(X,Y, ...)



```

**Comment by MLbyML** [comment]
Thank you Uwe @uschmidt83 for sharing the insight about how the StarDist datagenerator works. That helped!
Thanks Martin @maweigert for the code snippet - the idea completely makes sense. Will try it out and update how it goes!

**Comment by uschmidt83** [comment]
If that's slow, it might help to cache some of the loaded files:

```python
from functools import lru_cache
... 
    @lru_cache(50)
    def __getitem__(self, i):
        ...
```

**Comment by MLbyML** [comment]
Works very nicely. Minor extension to Martin and Uwe's code from above, but just for completion:

```from tensorflow.keras.utils import Sequence
import tifffile
from functools import lru_cache
 
class FileData(Sequence):
    def __init__(self, filenames, name, key = 'image', type= 'train'):
        super().__init__() 
        self._filenames= filenames
        self._key = key
        self._name = name
        print('{} {}-{} dataset created'.format(name, type, key))
        
    def __len__(self):
        return len(self._filenames)
    
    @lru_cache(100)
    def __getitem__(self, index):
        x = tifffile.imread(self._filenames[index])
        if self._key=='image':
            x = normalize(x, 1, 99.8, axis=(0,1,2))
        elif self._key=='mask':
            x = fill_label_holes(x)
        
        return x

X_trn = FileData(X_trn, name = 'organoid-2020', key ='image', type='train')
Y_trn = FileData(Y_trn, name = 'organoid-2020', key ='mask', type='train')
```
My problem is solved. Please feel free to close the issue. Thank you!

**Comment by lucas-sancere** [comment]
Hello Stardist team,

   Thank you for your work! Unfortunately, I still have a memory issue even using @maweigert proposal.
 
   Firstly, you can see just below the memory increase that I faced:


```
Epoch 20 Beginning

16:32:22 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      278769       45396         368       61033      103489
Swap:          8191        5230        2961


Epoch 20 End 

16:33:16 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      283010       41155         368       61033       99248
Swap:          8191        5230        2961


Epoch 25 Beginning

16:34:33 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      284364       39797         368       61038       97894
Swap:          8191        5230        2961


Epoch 25 End 

16:39:46 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      287205       36955         368       61038       95053
Swap:          8191        5228        2963

 

Epoch 30 Beginning

16:41:14 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      289252       34906         368       61041       93008
Swap:          8191        5228        2963


Epoch 30 End 

16:47:35 [lsancere@sulaco TensorFlowGPU]# free -m
              total        used        free      shared  buff/cache   available
Mem:         385200      293019       31139         368       61041       89240
Swap:          8191        5228        2963

```

Regarding what you proposed,  we add the following lines in our code to use a dynamic sequence:


```
ind = rng.permutation(len(Raw))
self.X= self.DataSequencer(Raw, self.axis_norm, Normalize = True, labelMe = False)
self.Y= self.DataSequencer(RealMask, self.axis_norm, Normalize = False, labelMe = True)

n_val = max(1, int(round(0.15 * len(ind))))
ind_train, ind_val = ind[:-n_val], ind[-n_val:]

self.X_val, self.Y_val = [self.X[i] for i in ind_val]  , [self.Y[i] for i in ind_val]
self.X_trn, self.Y_trn = [self.X[i] for i in ind_train], [self.Y[i] for i in ind_train]
``` 

With the following DataSequence class:
 
```
class DataSequencer(Sequence):
         
         
            def __init__(self, files, axis_norm, Normalize = True, labelMe = False):
                super().__init__() 
                
                self.files = files
               
                self.axis_norm = axis_norm
                self.labelMe = labelMe
                self.Normalize = Normalize
                
            def __len__(self):
                return len(self.files)
            
            
            def __getitem__(self, i):
                
                        #Read Raw images
                         if self.Normalize == True:
                                 x = ReadFloat(self.files[i]) 
                                 x = normalize(x,1,99.8,axis= self.axis_norm) 
                         if self.labelMe == True:
                                 #Read Label images
                                 x = ReadInt(self.files[i])
                                 x = label(x)
                   
                         return x
```

ReadFloat and ReadInt are the following:

```
def ReadFloat(fname):

    return imread(fname).astype('float16')         
         

def ReadInt(fname):

    return imread(fname).astype('uint16')  
```

The training images have size between 384 by 384 pixels to 4000 by 4000 pixels and we choose a patch size of 384 by 384 in out training with 128 Rays. Total number of Images = 4216
    
Do you have any clues about this memory increase?
      
Thank you for your support.

**Comment by uschmidt83** [comment]
Hi @lucas-sancere,

is this actually a problem for you, or are you just curious?

> Additionally, the StarDist data generator makes an internal copy to cache regions where it can sample patches from. There are probably a few other places where copies of the data are needed...

The data generator [keeps a cache](https://github.com/mpicbg-csbd/stardist/blob/0c737544a05ae3f6b60667516d2d25e0cebef265/stardist/models/base.py#L130-L145) which will grow during training. Maybe that's the reason?

Best,
Uwe

**Comment by lucas-sancere** [comment]
Hi @uschmidt83

No, it is not a problem finally. I thought it was an issue, but finally the increase stopped before reaching memory limit. Thanks, it should be this, I will try to see. 

Thank you for your support, 
Regards

**Comment by maweigert** [comment]
Adding for completeness:

We added the option to disable the sample region cache that is normally keeps on building during training (currently in `dev`). 

So the best approach for large datasets that don't fit into memory is currently:

* Install the `dev` branch (when using `stardist <= 0.6.2`)
*  Use a `keras.Sequence` to lazily load the data like in [this example above](https://github.com/stardist/stardist/issues/107#issuecomment-750215625)
* Disable sample cache in the config `config = Config3D(..., train_sample_cache = False)`

**Comment by kapoorlab** [comment]
Thanks @maweigert,
I did the lazy load as you mention in the example, I took a very small subset of my dataset, 10 ZYX image pairs and after waiting for the lazy load I got this error at the start of the training and then nothing happens after that. If we could solve this issue I think we are good to go.
`Epoch 1/200
data_utils.py (718): An input could not be retrieved. It could be because a worker has died.We do not have any information on the lost sample.`
